internal/daemon/cache/repository_token.go

```diff
@@ -6,125 +6,223 @@ package cache
-			}
-			onConflict := &db.OnConflict{
-				Target: db.Columns{"id"},
-				Action: db.DoNothing(true),
-			}
-			if err := writer.Create(ctx, u, db.WithOnConflict(onConflict)); err != nil {
-				return errors.Wrap(ctx, err, op)
+		for _, at := range ats {
+			if oldestUsersTime == nil || oldestUsersTime.After(at.LastAccessedTime) {
+				oldestUser = u
+				oldestUsersTime = &at.LastAccessedTime
 			}
 		}
-
-		{
-			st := &AuthToken{
-				Id:               at.Id,
-				UserId:           at.UserId,
-				LastAccessedTime: time.Now(),
-			}
-			onConflict := &db.OnConflict{
-				Target: db.Columns{"id"},
-				Action: db.SetColumns([]string{"last_accessed_time"}),
-			}
-			if err := writer.Create(ctx, st, db.WithOnConflict(onConflict)); err != nil {
-				return errors.Wrap(ctx, err, op)
-			}
+	}
+	if oldestUser != nil {
+		if _, err := deleteUser(ctx, writer, oldestUser); err != nil {
+			return errors.Wrap(ctx, err, op)
 		}
+	}
+	return nil
+}
 
-		{
-			onConflict := &db.OnConflict{
-				Target: db.Columns{"keyring_type", "token_name"},
-				Action: db.SetColumns([]string{"auth_token_id"}),
+// AddRawToken upserts the auth token's user and auth token in the db and
+// stores the actual raw auth token in the repositories in memory storage.
+// The raw token must be valid and present in boundary and be for a user that
+// has permission to send a Read request for itself to boundary.
+func (r *Repository) AddRawToken(ctx context.Context, bAddr string, rawToken string) error {
+	const op = "cache.(Repository).AddRawToken"
+	switch {
+	case rawToken == "":
+		return errors.New(ctx, errors.InvalidParameter, op, "boundary auth token is empty")
+	case bAddr == "":
+		return errors.New(ctx, errors.InvalidParameter, op, "boundary address is empty")
+	}
+
+	atIdParts := strings.SplitN(rawToken, "_", 4)
+	if len(atIdParts) != 3 {
+		return errors.New(ctx, errors.InvalidParameter, op, "boundary auth token is is malformed")
+	}
+	atId := strings.Join(atIdParts[:2], "_")
+
+	var at *authtokens.AuthToken
+	{
+		var inMemAuthToken *authtokens.AuthToken
+		atV, inMem := r.idToKeyringlessAuthToken.Load(atId)
+		if inMem {
+			var ok bool
+			inMemAuthToken, ok = atV.(*authtokens.AuthToken)
+			if !ok {
+				return errors.New(ctx, errors.Internal, op, "unable to cast in memory auth token to *authtoken.AuthToken")
 			}
-			if err := writer.Create(ctx, kt, db.WithOnConflict(onConflict)); err != nil {
+		}
+		t := &AuthToken{
+			Id: atId,
+		}
+		err := r.rw.LookupById(ctx, t)
+		switch {
+		case err != nil && !errors.IsNotFoundError(err):
+			return errors.Wrap(ctx, err, op)
+		case errors.IsNotFoundError(err) || !inMem:
+			// if we don't know about it in the cache or we don't know about
+			// this auth token in memory, get it from boundary to sure up the
+			// cache information about this auth token.
+			at, err = r.tokenReadFn(ctx, bAddr, rawToken)
+			if err != nil {
 				return errors.Wrap(ctx, err, op)
 			}
+		case t.UserId != inMemAuthToken.UserId:
```

Comment:
This is part of the contract, so perhaps we should have a unit test for when it
doesn't match?